sdk-lib/src/main/rust/utils/trace.rs

```diff
@@ -0,0 +1,46 @@
+use std::ffi::CString;
+
+use tracing::{error, span, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan};
+
+use super::target_ndk::{Api23, NdkApi};
+
+pub struct Layer {
+    ndk_api: Option<NdkApi>,
+}
+
+impl Layer {
+    pub fn new(ndk_api: Option<NdkApi>) -> Self {
+        Layer { ndk_api }
+    }
+
+    fn with_api(&self, f: impl FnOnce(&Api23)) {
+        if let Some(v23) = self.ndk_api.as_ref().and_then(|api| api.v23.as_ref()) {
+            f(v23)
+        }
+    }
+}
+
+impl<S: Subscriber> tracing_subscriber::Layer<S> for Layer
+where
+    for<'lookup> S: LookupSpan<'lookup>,
+{
+    fn on_enter(&self, id: &span::Id, ctx: Context<'_, S>) {
+        self.with_api(|api| match ctx.metadata(id) {
+            Some(metadata) => match CString::new(metadata.name()) {
+                Ok(section_name) => unsafe { api.ATrace_beginSection(section_name.as_ptr()) },
+                Err(_) => error!(
+                    "Span name contains internal NUL byte: '{}'",
+                    metadata.name()
+                ),
+            },
+            None => error!("Span {:?} has no metadata", id),
+        });
+    }
+
+    fn on_exit(&self, _id: &span::Id, _ctx: Context<'_, S>) {
+        self.with_api(|api| {
+            unsafe { api.ATrace_endSection() };
```

Comment:
If the span name had an internal NUL then `ATrace_endSection` will be called without a corresponding `ATrace_beginSection`. The [NDK docs](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1gacd849f3cdbe927673c0ad8740c369aed) say:
> `void ATrace_endSection()`
>
> Writes a tracing message to indicate that a given section of code has ended.
>
> This call must be preceeded by a corresponding call to [ATrace_beginSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1ga13768bfdb5d90166ec50bc41235e2797) on the same thread. Calling this method will mark the end of the most recently begun section of code, so care must be taken to ensure that [ATrace_beginSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1ga13768bfdb5d90166ec50bc41235e2797)/[ATrace_endSection](https://developer.android.com/ndk/reference/group/tracing#group___tracing_1gacd849f3cdbe927673c0ad8740c369aed) pairs are properly nested and called from the same thread.

Reply:
There are similar issues if the Rust code's spans are across async Rust `.await` boundaries, which we can't deal with until API 29. I will be handling that in a separate PR.